src/ast/passes/semantic_analyser.cpp

```diff
@@ -1914,6 +1915,127 @@ void SemanticAnalyser::visit(While &while_block)
   loop_depth_--;
 }
 
+void SemanticAnalyser::visit(For &f)
+{
+  bool error = false;
+
+  if (!bpftrace_.feature_->has_helper_for_each_map_elem()) {
+    LOG(ERROR, f.loc, err_)
+        << "Missing required kernel feature: for_each_map_elem";
+    error = true;
+  }
+
+  /*
+   * For-loops are implemented using the bpf_for_each_map_elem helper function,
+   * which requires them to be rewritten into a callback style.
+   *
+   * In this first implementation, we do not allow any variables to be shared
+   * between the main probe and the loop's body. Maps are global so are not a
+   * problem.
+   *
+   * Pseudo code for the transformation we apply:
+   *
+   * Before:
+   *     PROBE {
+   *       @map[0] = 1;
+   *       for ($kv : @map) {
+   *         [LOOP BODY]
+   *       }
+   *     }
+   *
+   * After:
+   *     PROBE {
+   *       @map[0] = 1;
+   *       bpf_for_each_map_elem(@map, &map_for_each_cb, 0, 0);
+   *     }
+   *     long map_for_each_cb(bpf_map *map,
+   *                          const void *key,
+   *                          void *value,
+   *                          void *ctx) {
+   *       $kv = ((uint64)key, (uint64)value);
+   *       [LOOP BODY]
+   *     }
+   */
+
+  // Validate decl
+  const auto &decl_name = f.decl->ident;
+  if (variable_val_[probe_].find(decl_name) != variable_val_[probe_].end()) {
+    LOG(ERROR, f.decl->loc, err_)
+        << "Loop declaration shadows existing variable: " + decl_name;
+    error = true;
+  }
+
+  // Validate expr
+  if (!f.expr->is_map) {
+    LOG(ERROR, f.expr->loc, err_) << "Loop expression must be a map";
+    error = true;
+  }
+  Map &map = static_cast<Map &>(*f.expr);
+
+  // Validate body
+  CollectNodes<Variable> vars_referenced;
+  for (auto *stmt : *f.stmts) {
+    vars_referenced.run(*stmt);
+  }
+  for (const Variable &var : vars_referenced.nodes()) {
+    if (variable_val_[probe_].find(var.ident) != variable_val_[probe_].end()) {
+      LOG(ERROR, var.loc, err_) << "Variables defined outside of a for-loop "
+                                   "can not be accessed in the loop's scope";
+      error = true;
+    }
+  }
+
+  // This could be relaxed in the future:
+  CollectNodes<Jump> jumps;
+  for (auto *stmt : *f.stmts) {
+    jumps.run(*stmt);
+  }
+  for (const Jump &n : jumps.nodes()) {
+    LOG(ERROR, n.loc, err_)
+        << "'" << opstr(n) << "' statement is not allowed in a for-loop";
+    error = true;
+  }
+
+  if (error)
+    return;
```

Comment:
Nit: you could do the same as semantic analyser does at the end:
```suggestion
  if (!err_.empty())
    return;
```
That would spare the `error` variable.

Maybe we could even introduce some wrapper like `has_error` as it may be useful in other places, too.